woocommerce/woocommerce-ios
# Enable alternative types for certain fields when decoding a `ProductVariation`

Fixes #2401 

Note: ~304 diffs of this PR are from JSON files used to test decoding a product or product variation in different edge cases!

In addition to fixing the issue when a site's `ProductVariation` might have different types from expected, this PR also includes some refactoring on how alternative types are decoded.

## Changes

All the changes are in the Networking layer:

- Created `AlternativeDecodingType` enum with 3 types that are currently used:
  - `decimal`: used to try decoding a value of `Decimal` type
  - `string`: used to try decoding a value of `String` type
  - `bool`: used to try decoding a value of `Bool` type
- In `KeyedDecodingContainer`'s extension, added another helper `failsafeDecodeIfPresent` that takes a target type for the final type with a list of alternative types with how each alternative type is transformed to the target type
- In `Product`, updated the alternative types conditional logic with the new `failsafeDecodeIfPresent` helper with alternative types
  - Right now, we support alternative types for `price`, `salePrice`, and `manageStock`
  - Added a test case with a JSON that contains the fields with alternative types
- In `ProductVariation`, added/updated the alternative types conditional logic with the new `failsafeDecodeIfPresent` helper with alternative types
  - Similar to `Product`, we support alternative types for `price`, `salePrice`, and `manageStock`
  - Added test cases with JSON files for decoding a list of `ProductVariation`s with `ProductVariationListMapper`

## Testing

Fortunately, I was able to repro this on @Garance91540's site (thanks to Rachel's investigation in p5T066-1o1#comment-5400-p2)! There is probably a third-party plugin that alters the price type from string to decimal. However, feel free to test this on any site.

- Launch the app
- Go to the products tab --> the product list should load as before
- Tap on a variable product
- Tap on the variations row --> the product variation list should load correctly

## Example screenshots

`develop` | PR branch
-- | --
![Simulator Screen Shot - iPhone 11 - 2020-07-17 at 18 26 10](https://user-images.githubusercontent.com/1945542/87776997-45644600-c85b-11ea-8287-72d0b8fc29de.png) | ![Simulator Screen Shot - iPhone 11 - 2020-07-17 at 18 28 05](https://user-images.githubusercontent.com/1945542/87777007-485f3680-c85b-11ea-9757-6de6e4b107cd.png)


Update release notes:
